### DISCUSSION TITLE
How to not treat a str as a Vec<String>

### DISCUSSION THREAD

**Discussion by atoav** [PyO3/pyo3#2341] [Questions]
In python one can iterate over a `str` as if it was a `List[str]`. And while this is occasionaly useful it can also be a major footgun if users pass in a `"string"` instead of a `["string"]` and suddenly have a function operate on single characters instead of the whole string, as they might (admittedly, errouniously) assumed one that one string.

In python I'd prevent this user error  by ensuring incoming str are never treated as a list of characters by wrapping a lone `str` in a list as follows:

```python
if isinstance(strings, str):
    strings = [strings]
```

In pyo3 this is an issue as well. Let's say i have a simple function that should take a `Vec<String>` and print out each string in the vec:

```Rust
/// Print out each string in a list of strings
#[pyfunction]
fn print_strings(strings: Vec<String>) {
    for s in strings {
        println!("{}", s);
    }
}
```

If I now run this in the python interpreter with a string that is not a list each character is printed out seperately:  

```python
>>> from my_module import print_strings
>>> print_strings("string")
s
t
r
i
n
g
```

Is there any simple way to ensure single `str` are treated as `[str]` using pyo3? Even an Error would be acceptable, as long as users don't pass a string and have it iterate over characters without them noticing.

**Comment by davidhewitt** [comment]
I remember `pybind11` changed back in their 2.3.0 release to not cast `str` to `std::vector<std::string>`, I think actually we should do similar if we can.

**Reply by atoav** [reply]
Sounds like a good idea, because one could reasonably expect a stricter typecheck here (at least I did).
